Frontend atualizar saldo com base em transações recebidas

GCastilho opened the pull request:
-- O frontend irá receber eventos de new_transaction e update_received_tx e atualizar corretamente o saldo de acordo com eles
-- A tabela de balances irá mostrar uma coluna com o saldo 'locked', atualmente só mostra o available
-- O saldo client-side deve ser corretamente atualizado de acordo com os eventos recebidos

GCastilho committed:
CurrencyApi agora emite update_received_tx

-- A CurrencyApi agora emite um evento update_received_tx, com userId, currency e UpdtReceived como argumentos
-- Resolvido um bug em que o new_transaction estava sendo emitido com a interface errada
-- Refatorado os listeners internos da currencyApi para ficarem no contrutor
-- Corrigido um bug no emitter do socket que não estava emitindo a currency no evento de new_transaction
-- Adicionado um emitter para update_received_tx, que emite um evento com currency e txUpdate para o usuário, caso ele esteja conectado

joaojvds committed:
Acrescentado balance locked na tabela

--Alterado balance para balance available

joaojvds committed:
Amount fix

--Mudado o amount de string para number

GCastilho committed:
Resolvido um bug no websocket

-- Resolvido um bug em que a connectedUsers estava usando objetos como chaves do MAP
-- Resolvido um bug onde a emitters não estava sendo importada

joaojvds committed:
Valor atualizado ao receber nova transação

--Ao receber o evento new_transaction, o balance locked atualiza o seu valor

joaojvds committed:
Adicionado handler para update_received_tx

--Ao receber um evento com a confirmação da transaction, ele atualiza os valores no site (Precisa ser testado)

joaojvds committed:
Removido redundância do código

GCastilho commented on client/src/stores/balances.js:
```diff
@@ -21,3 +21,35 @@ auth.subscribe(async auth => {
+ */
+addSocketListener('new_transaction', (currency, transaction) => {
+	update(balances => {
+		balances[currency].locked += transaction.amount
```
Uma transação da NANO emite o evento new_transaction com status 'confirmed', então nem sempre é só o locked que vc tem que incrementar, isso depende da prop 'status' da transaction

GCastilho reviewed (commented):
Um detalhe q eu não tinha percebido da última vez

Se vc olhar a interface do TxReceived, tem que o status dela pode ser 'pending' ou 'confirmed', vc está presumindo que vai ser sempre pending (no bitcoin, TENDE A SER, mas tbm não é garantido)

GCastilho committed:
Merge branch 'master' into frontend-new-tx

-- As recentes atualizações no master são necessárias para o desenvolvimento desse branch

GCastilho committed:
Adicionado retorno to get_tx_info

-- O request 'get_opid_info' do cliente foi modificado para 'get_tx_info'
-- Implementado uma interface TxInfo na Transactions para padronizar o retorno de informações de transações
-- Implementado um listener global do backend para 'get_tx_info', que retorna um TxInfo ao cliente caso esta seja dele

joaojvds commented:
Ao chamar o evento get_tx_info, ele da o erro NotAuthorized

joaojvds reviewed (commented):
Ao chamar o evento get_tx_info, ele da o erro NotAuthorized

joaojvds committed:
Alterações nos handlers 'new_transaction' e 'update_received_tx'

--Salvando transações ao receber o evento new_transaction
--Alterando o amount available ao receber o evento new_transaction com o status 'confirmed'
--Quando disponivel, ele irá usar os dados da transaction salvas

GCastilho commented on client/src/stores/balances.js:
```diff
@@ -1,9 +1,11 @@
 
 const { subscribe, set, update } = writable({})
 
+let transactions = []
```
Isso deve ficar em uma store/módulo separado, pois não é só o balances que irá precisar dele

GCastilho commented on client/src/stores/balances.js:
```diff
@@ -21,3 +23,49 @@ auth.subscribe(async auth => {
+addSocketListener('update_received_tx', async (currency, txUpdate) => {
+	console.log(txUpdate)
+	if (txUpdate.status === 'confirmed') {
+		if (transactions.some(transaction => transaction.opid === txUpdate.opid)) {
```
Você está varrendo o array a cada atualização dele, isso é extremamente ineficiente, não tem uma maneira melhor de salvar as transações? Tipo um Map, pra não ter que varrer um array a cada update?

GCastilho commented on client/src/stores/balances.js:
```diff
@@ -21,3 +23,49 @@ auth.subscribe(async auth => {
+	if (txUpdate.status === 'confirmed') {
+		if (transactions.some(transaction => transaction.opid === txUpdate.opid)) {
+			update(balances => {
+				const txInfo = transactions.filter(transaction => transaction.opid === txUpdate.opid)
```
E agora vc está varrendo o array uma segunda vez para pegar o dado que vc quer

Pq fazer isso duas vezes?

GCastilho reviewed (commented):
Dê uma pesquisada _se é possível_ salvar as transações em um Map, pq é muito mais eficiente que varrer array

GCastilho committed:
Resolvido bug do "NotAuthorized"

-- Resolvido bug de todas as solicitações por "get_tx_info" estarem dando "NotAuthorized"

joaojvds committed:
Criado store das transactions

--Criado função para pegar transação pelo opid

joaojvds committed:
quick fix

GCastilho commented on client/src/stores/transactions.js:
```diff
@@ -0,0 +1,14 @@
+	let transaction
+	subscribe(transactions => {
+		transaction = transactions.find(tx => tx.opid === opid)
+	})
```
Tem um memory leak aqui

Você está fazendo um subscribe e não está fazendo o unsubscribe

Uma nota: Fazer o unsubscribe logo em seguida são duas function calls uma após a outra. O que, entre outras coisas, gasta recursos

GCastilho commented on client/src/stores/balances.js:
```diff
@@ -21,3 +22,52 @@ auth.subscribe(async auth => {
+				/** Usa dados do amount pego no 'txInfo' para atualizar o saldo na tela */
+				update(balances => {
+					balances[currency].available += +txInfo.amount.$numberDecimal
+					balances[currency].locked -= +txInfo.amount.$numberDecimal
```
WTF?

joaojvds replied:
esse é o retorno do get_tx_info 
`{ 
     amount: {
        $numberDecimal: "2e-7"
     }
}`

GCastilho replied:
Mas ele deveria retornar um number, se n tá retornando é um bug

Pq vc não comentou disso?

joaojvds replied:
Dessa vez eu esqueci mesmo

joaojvds replied:
Quando o bug for consertado eu vou tirar esse segundo update, já que ele esta ai só por conta disso

GCastilho replied:
ok

GCastilho replied:
Fixed!

GCastilho commented:
aaaai carai...

Vou corrigir isso daqui a pouco, já pode presumir que vai ser um
number daqui em diante

2020-02-25 22:55 GMT-03:00, João Victor <notifications@github.com>:
> joaojvds commented on this pull request.
>
>
>
>> +					balances[currency].available += +txInfo.amount.$numberDecimal
> +					balances[currency].locked -= +txInfo.amount.$numberDecimal
>
> Dessa vez eu esqueci mesmo
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/GCastilho/unip-tcc/pull/37#discussion_r384235680

GCastilho commented:
@joaojvds  Não dá pra store de transactions ser um Map?

joaojvds commented:
@GCastilho Eu acho que é mais simples nesse caso trabalhar com arrays, apesar de que se eu colocar o opid como index do map seja mais facil pegar o valor já que eu só usaria `Map.get(opid)` para pegar os dados, mas as transações ainda não estão com a propriedade opid então não tem muito o que eu fazer

GCastilho commented:
Como assim elas não tem o opid? O opid n vem com o new_transaction?

joaojvds commented:
Eu avisei você e você disse para eu presumir que vem e continuar trabalhando, achei que você estava trabalhando nisso

GCastilho commented:
Eu estava, daí eu terminei e mandei. Achei q tava claro essa parte

Comunicação é uma coisa muito importante

joaojvds committed:
Resolvido memory leak

--Resolvido memory leak em em transactions.js
--Colocado emit 'get_tx_info' em transactions.js

GCastilho commented on client/src/stores/transactions.js:
```diff
@@ -0,0 +1,28 @@
+	let transaction
+	const unsubscribe = subscribe(transactions => {
+		transaction = transactions.find(tx => tx.opid === opid)
+	})
```
Eu não gosto desse unsubscribe, mas eu não tenho uma ideia melhor, então deixa assim por enquanto

GCastilho replied:
Tive uma ideia melhor :D

GCastilho commented on client/src/stores/transactions.js:
```diff
@@ -0,0 +1,28 @@
+			console.error('Error while getting transaction info:', err)
+		}
+	}
+	onDestroy(unsubscribe)
```
Isso aqui não funciona pq o onDestroy nunca vai ser chamado pq esse módulo não é um componente, e ele só funciona em componentes

joaojvds replied:
Nesse caso o que eu uso?

GCastilho replied:
Acho que a única opção é usar o `unsubscribe()` cada vez q roda a função

Mas isso é mto feio (e pesado), aff

Vou ver se penso em alguma coisa

GCastilho replied:
Fixed!

GCastilho reviewed (commented):
Finalmente vc teve a ideia de colocar o pedido por informações da transação no módulo da transação, tava pensando em quando vc teria essa ideia

GCastilho committed:
Amount agora vem em number

-- Corrigido bug na 'get_tx_info' em que o amount estava vindo em Decimal128 ao invés de number

joaojvds commented:
@GCastilho  Dessa vez você colocou o opid no retorno do new_transaction ?

GCastilho commented:
Como assim? não existe retorno do new_transaction

joaojvds commented:
Eu já tinha te avisado, o new_transaction não vem a propriedade opid

GCastilho commented:
É, eu acabei de reparar isso. É um bug, estou trabalhando pra consertar ele

GCastilho committed:
Merge branch 'master' into frontend-new-tx

GCastilho committed:
Corrigido bug do opid não estar sendo enviado na new_transaction

-- A new_transaction agora envia um objeto do tipo 'TxInfo' ao invés de um 'TxReceived'
-- Adicionado uma interface para padronizar os eventos públicos da Common e da CurrencyApi

GCastilho committed:
Merge branch 'master' into frontend-new-tx

GCastilho committed:
Reprojetado o sistema de armazenamento de transações

-- Transações agora são salvas em um módulo 'transactions' ao invés de uma store
-- Simplificado o funcionamento da store 'balances'

GCastilho committed:
Corrigido o código de acordo com as novas regras do eslint

-- Corrigido o código de acordo com as novas regras do eslint

GCastilho commented:
O que mais falta se implementado?

joaojvds commented:
Falta resolver o memory leak, e testar para ver se esta tudo funcionando, o problema é que demora uma eternidade para receber o confirmed, para mim leva mais de 40 min para confirmar

GCastilho commented:
Eu já resolvi o memory leak reprojetando como o armazenamento de
transações funciona

Sobre a demora, é, bitcoin é assim, bem vindo xD

GCastilho committed:
Movido websocket para utils

-- Movido módulo do websocket para a pasta utils

GCastilho committed:
Servidor agora emite 'update_sent_tx'

-- O servidor agora emite o evento update_sent_tx aos usuários
-- Leve limpeza no código, reduzindo types inúteis e comentários irrelevantes

joaojvds committed:
Criado handler para update_sent_tx

--Criado handler para atualizar o balance ao receber confirmação

GCastilho commented on client/src/stores/balances.js:
```diff
@@ -18,6 +19,55 @@ auth.subscribe(async auth => {
+ * Atualiza o balance locked ao confirmar envio da transação
+ */
+addSocketListener('update_sent_tx', async (currency, txSent) => {
+	console.log(txSent)
```
Deixa esse log mais claro, igual o da received

GCastilho commented on client/src/stores/balances.js:
```diff
@@ -18,6 +19,55 @@ auth.subscribe(async auth => {
+	try {
+		const txInfo = await transactions.getByOpid(txSent.opid)
+		update(balances => {
+			balances[currency].locked -= +txInfo.amount
```
Não precisa de cast to number numa subtração

GCastilho committed:
Set de transações é atualizado com evento de update

-- Adicionado listeners no módulo de transactions para atualizar dados locais de uma transação salva
-- Minor fixes

joaojvds committed:
Melhorado mensagem de log do update_sent_tx

GCastilho committed:
Minor fix

-- Trocado Balance locked/available por Available/Locked balance
